README
NAME

    Rewire - Dependency Injection

ABSTRACT

    Dependency Injection Container for Perl 5

SYNOPSIS

      use Rewire;
    
      my $services = {
        filetemp => {
          package => 'File/Temp'
        },
        tempfile => {
          package => 'Mojo/File',
          argument => { '$service' => 'filetemp' }
        }
      };
    
      my $rewire = Rewire->new(services => $services);
    
      $rewire->resolve('tempfile');

DESCRIPTION

    This package provides methods for using dependency injection, and
    building objects and values.

INTEGRATES

    This package integrates behaviors from:

    Data::Object::Role::Buildable

    Data::Object::Role::Proxyable

LIBRARIES

    This package uses type constraints from:

    Types::Standard

SCENARIOS

    This package supports the following scenarios:

 $envvar

      use Rewire;
    
      my $services = {
        file => {
          package => 'Mojo/File',
          argument => { '$envvar' => 'home' }
        }
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports inlining environment variables as arguments to
    services. The $envvar directive is used to specify the name of an
    environment variable, and can also be used in metadata for reusability.

 $function

      use Rewire;
    
      my $services = {
        temp => {
          package => 'File/Temp'
        },
        file => {
          package => 'Mojo/File',
          argument => { '$function' => 'temp#tempfile' }
        }
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports inlining the result of a service resolution and
    function call as arguments to services. The # delimited $function
    directive is used to specify the name of an existing service on the
    right-hand side, and an arbitrary function to be call on the result on
    the left-hand side.

 $metadata

      use Rewire;
    
      my $metadata = {
        home => '/home/ubuntu'
      };
    
      my $services = {
        file => {
          package => 'Mojo/File',
          argument => { '$metadata' => 'home' }
        }
      };
    
      my $rewire = Rewire->new(
        metadata => $metadata,
        services => $services
      );

    This package supports inlining configuration data as arguments to
    services. The $metadata directive is used to specify the name of a
    stashed configuration value or data structure.

 $method

      use Rewire;
    
      my $services = {
        temp => {
          package => 'File/Temp'
        },
        file => {
          package => 'Mojo/File',
          argument => { '$method' => 'temp#filename' }
        }
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports inlining the result of a service resolution and
    method call as arguments to services. The # delimited $method directive
    is used to specify the name of an existing service on the right-hand
    side, and an arbitrary method to be call on the result on the left-hand
    side.

 $routine

      use Rewire;
    
      my $services = {
        temp => {
          package => 'File/Temp'
        },
        file => {
          package => 'Mojo/File',
          argument => { '$routine' => 'temp#tempfile' }
        }
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports inlining the result of a service resolution and
    routine call as arguments to services. The # delimited $routine
    directive is used to specify the name of an existing service on the
    right-hand side, and an arbitrary routine to be call on the result on
    the left-hand side.

 $service

      use Rewire;
    
      my $services = {
        io => {
          package => 'IO/Handle'
        },
        log => {
          package => 'Mojo/Log',
          argument => {
            handle => { '$service' => 'io' }
          }
        },
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports inlining resolved services as arguments to other
    services. The $service directive is used to specify the name of a
    service to be resolved and passed as an argument.

 arguments

      use Rewire;
    
      my $metadata = {
        applog => '/var/log/rewire.log'
      };
    
      my $services = {
        mojo_log => {
          package => 'Mojo/Log',
          argument => {
            path => { '$metadata' => 'applog' },
            level => 'warn'
          },
          argument_as => 'list'
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
        metadata => $metadata
      );

    This package supports providing static and/or dynamic arguments during
    object construction from metadata or other services.

 builder

      use Rewire;
    
      my $services = {
        mojo_date => {
          package => 'Mojo/Date',
          builder => [
            {
              method => 'new',
              return => 'self'
            },
            {
              method => 'to_datetime',
              return => 'result'
            }
          ]
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
      );

    This package supports specifying multiple build steps as function,
    method, and routine calls and chaining them together.

 config

      use Rewire;
    
      my $metadata = {
        home => '/home/ubuntu'
      };
    
      my $services = {
        tempfile => {
          package => 'Mojo/File',
          argument => { '$metadata' => 'home' }
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
        metadata => $metadata
      );

    This package supports configuring services and metadata in the service
    of building objects and values.

 constructor

      use Rewire;
    
      my $services = {
        mojo_date => {
          package => 'Mojo/Date',
          constructor => 'new'
        }
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports specifying constructors other than the
    traditional new routine. A constructor is always called with the
    package name as the invocant.

 extends

      use Rewire;
    
      my $services = {
        io => {
          package => 'IO/Handle'
        },
        log => {
          package => 'Mojo/Log',
          argument => {
            handle => { '$service' => 'io' }
          }
        },
        development_log => {
          package => 'Mojo/Log',
          extends => 'log',
          builder => [
            {
              method => 'path',
              argument => '/tmp/development.log',
              return => 'none'
            },
            {
              method => 'level',
              argument => 'debug',
              return => 'none'
            }
          ]
        },
        production_log => {
          package => 'Mojo/Log',
          extends => 'log',
          builder => [
            {
              method => 'path',
              argument => '/tmp/production.log',
              return => 'none'
            },
            {
              method => 'level',
              argument => 'warn',
              return => 'none'
            }
          ]
        },
      };
    
      my $rewire = Rewire->new(
        services => $services
      );

    This package supports extending services in the definition of other
    services, effectively using the extended service as the invocant in the
    creation of the requested service.

 function

      use Rewire;
    
      my $services = {
        foo_sum => {
          package => 'Mojo/Util',
          function => 'md5_sum',
          argument => 'foo',
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
      );

    This package supports specifying construction as a function call, which
    when called does not provide an invocant.

 lifecycle

      use Rewire;
    
      my $metadata = {
        home => '/home/ubuntu'
      };
    
      my $services = {
        tempfile => {
          package => 'Mojo/File',
          argument => { '$metadata' => 'home' },
          lifecycle => 'singleton'
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
        metadata => $metadata
      );

    This package supports different lifecycle options which determine when
    services are built and whether they're persisted.

 metadata

      use Rewire;
    
      my $metadata = {
        homedir => '/home',
        tempdir => '/tmp'
      };
    
      my $services = {
        home => {
          package => 'Mojo/Path',
          argument => { '$metadata' => 'homedir' },
        },
        temp => {
          package => 'Mojo/Path',
          argument => { '$metadata' => 'tempdir' },
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
        metadata => $metadata
      );

    This package supports specifying data and structures which can be used
    in the construction of multiple services.

 method

      use Rewire;
    
      my $services = {
        mojo_url => {
          package => 'Mojo/URL',
          argument => 'https://perl.org',
          method => 'new'
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
      );

    This package supports specifying construction as a method call, which
    when called provides the package or object instance as the invocant.

 proxyable

      use Rewire;
    
      my $metadata = {
        homedir => '/home',
        tempdir => '/tmp'
      };
    
      my $services = {
        home => {
          package => 'Mojo/Path',
          argument => { '$metadata' => 'homedir' },
        },
        temp => {
          package => 'Mojo/Path',
          argument => { '$metadata' => 'tempdir' },
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
        metadata => $metadata
      );

    This package supports the resolution of services using a single method
    call. This is enabled by intercepting method calls and proxying them to
    the "process" method.

 routine

      use Rewire;
    
      my $services = {
        mojo_url => {
          package => 'Mojo/URL',
          argument => 'https://perl.org',
          routine => 'new'
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
      );

    This package supports specifying construction as a function call, which
    when called provides the package as the invocant.

 service

      my $metadata = {
        home => '/home/ubuntu'
      };
    
      my $services = {
        tempfile => {
          package => 'Mojo/File',
          argument => { '$metadata' => 'home' },
          lifecycle => 'eager'
        }
      };
    
      my $rewire = Rewire->new(
        services => $services,
        metadata => $metadata
      );

    This package supports defining services to be constructed on-demand or
    automatically on instantiation.

ATTRIBUTES

    This package has the following attributes:

 context

      context(CodeRef)

    This attribute is read-only, accepts (CodeRef) values, and is optional.

 engine

      engine(InstanceOf["Data::Object::Space"])

    This attribute is read-only, accepts
    (InstanceOf["Data::Object::Space"]) values, and is optional.

 metadata

      metadata(HashRef)

    This attribute is read-only, accepts (HashRef) values, and is optional.

 services

      services(HashRef)

    This attribute is read-only, accepts (HashRef) values, and is optional.

METHODS

    This package implements the following methods:

 config

      config() : HashRef

    The config method returns the configuration based on the services and
    metadata attributes.

    config example #1

        # given: synopsis
      
        $rewire->config;

 process

      process(Str $name, Any $argument, Maybe[Str] $argument_as) : Any

    The process method processes and returns an object or value based on
    the service named but where the arguments are provided ad-hoc. Note:
    This method is meant to be used to construct services ad-hoc and as
    such bypasses caching and lifecycle effects.

    process example #1

        # given: synopsis
      
        $rewire->process('tempfile', 'rewire.tmp');

    process example #2

        use Rewire;
      
        my $metadata = {
          logfile => '/var/log/rewire.log',
        };
      
        my $services = {
          mojo_log => {
            package => 'Mojo/Log',
            argument => { '$metadata' => 'logfile' },
          }
        };
      
        my $rewire = Rewire->new(
          services => $services,
          metadata => $metadata
        );
      
        $rewire->process('mojo_log', {
          level => 'fatal',
          path => { '$metadata' => 'logfile' }
        });

    process example #3

        use Rewire;
      
        my $metadata = {
          logfile => '/var/log/rewire.log',
        };
      
        my $services = {
          mojo_log => {
            package => 'Mojo/Log',
            builder => [
              {
                method => 'new',
                return => 'self'
              }
            ]
          }
        };
      
        my $rewire = Rewire->new(
          services => $services,
          metadata => $metadata
        );
      
        $rewire->process('mojo_log', {
          level => 'fatal',
          path => { '$metadata' => 'logfile' }
        });

 resolve

      resolve(Str $name) : Any

    The resolve method resolves and returns an object or value based on the
    service named. Note: This method is recommended to be used to construct
    services as defined by the configuration and as such doesn't not allow
    passing additional arguments.

    resolve example #1

        # given: synopsis
      
        $rewire->resolve('tempfile');

    resolve example #2

        use Rewire;
      
        my $services = {
          mojo_log => {
            package => 'Mojo/Log',
            argument => {
              level => 'fatal',
              path => '/var/log/rewire.log'
            },
          }
        };
      
        my $rewire = Rewire->new(
          services => $services,
        );
      
        $rewire->resolve('mojo_log');

    resolve example #3

        package Dynamic;
      
        sub import;
      
        sub AUTOLOAD {
          bless {};
        }
      
        sub DESTROY {
          ; # noop
        }
      
        package main;
      
        use Rewire;
      
        my $services = {
          dynamic => {
            package => 'Dynamic',
            builder => [
              {
                method => 'new',
                return => 'self'
              },
              {
                method => 'missing_method',
                return => 'result'
              }
            ],
          }
        };
      
        my $rewire = Rewire->new(
          services => $services,
        );
      
        $rewire->resolve('dynamic');

 validate

      validate() : Object

    The validate method validates the configuration and throws an exception
    if invalid, otherwise returns itself.

    validate example #1

        # given: synopsis
      
        $rewire->validate;

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the terms of the The Apache License, Version 2.0, as elucidated in the
    "license file"
    <https://github.com/iamalnewkirk/rewire/blob/master/LICENSE>.

PROJECT

    Wiki <https://github.com/iamalnewkirk/rewire/wiki>

    Project <https://github.com/iamalnewkirk/rewire>

    Initiatives <https://github.com/iamalnewkirk/rewire/projects>

    Milestones <https://github.com/iamalnewkirk/rewire/milestones>

    Contributing
    <https://github.com/iamalnewkirk/rewire/blob/master/CONTRIBUTE.md>

    Issues <https://github.com/iamalnewkirk/rewire/issues>



CHANGES
Changelog for Rewire

0.05 2020-05-20T01:50:33
 - Add release updates
 - Add proxyable services (#26)
 - Add support for packages with autoloading (#25)
 - Add support for ad-hoc results as arguments (#24)
 - Add support for extending services (#23)
 - Add support for passing args to final build step (#22)
 - Add support for using envvars in arguments (#21)

0.04 2020-05-14T17:40:14
 - Built release version 0.04
 - Add release updates
 - More documentation for "process" and "resolve" methods (#12)

0.03 2020-05-08T20:03:13
 - Built release version 0.03
 - Added release updates
 - Add VERSION token
 - Updated documentation
 - Fix "process" method w/single-key arg

0.02 2020-05-08T15:39:37
 - Built release version 0.02
 - Added release updates
 - Updated documentation
 - Add "process" method
 - Miscellaneous Housekeeping

0.01 2020-05-08T02:38:28
 - Built release version 0.01
 - Added release updates
 - Initial commit


lib/Rewire.pm
package Rewire;

use 5.014;

use strict;
use warnings;

use registry;
use routines;

use Data::Object::Class;
use Data::Object::ClassHas;
use Data::Object::Space;

use Carp;
use JSON::Validator;
use Rewire::Engine;

with 'Data::Object::Role::Buildable';
with 'Data::Object::Role::Proxyable';

# VERSION

# BUILD

fun build_self($self, $args) {

  # build context and eager load services
  return $self->context;
}

fun build_proxy($self, $package, $method, @args) {
  return unless $self->config->{services}{$method};

  return sub {

    return $self->process($method, @args);
  }
}

# ATTRIBUTES

has 'context' => (
  is => 'ro',
  isa => 'CodeRef',
  new => 1
);

fun new_context($self) {
  $self->engine->call('preload', $self->config);
}

has 'engine' => (
  is => 'ro',
  isa => 'InstanceOf["Data::Object::Space"]',
  new => 1
);

fun new_engine($self) {
  Data::Object::Space->new('Rewire::Engine');
}

has 'metadata' => (
  is => 'ro',
  isa => 'HashRef',
  opt => 1
);

has 'services' => (
  is => 'ro',
  isa => 'HashRef',
  opt => 1
);

# METHODS

method config() {
  {
    metadata => $self->metadata || {},
    services => $self->services || {},
  }
}

method resolve(Str $name) {
  my $engine = $self->engine;

  my $result = $engine->call('reifier', $name, $self->config, $self->context);

  return $result;
}

method process(Str $name, Any $argument, Maybe[Str] $argument_as) {
  my $engine = $self->engine;
  my $service = $self->services->{$name} or return;

  my $generated = {
    %$service, $argument_as ? (argument_as => $argument_as) : ()
  };

 $argument //= $service->{argument};

  my $params = $engine->call('resolver', $argument, $self->config, $self->context);
  my $result = $engine->call( 'builder', $generated, $params // $service->{argument});

  return $result;
}

method validate() {
  my $engine = $self->engine;

  my $json = JSON::Validator->new;

  $json->schema($engine->call('ruleset'));

  my @errors = map "$_", $json->validate($self->config);

  confess join "\n", @errors if @errors;

  return $self;
}

1;


lib/Rewire/Engine.pm
package Rewire::Engine;

use 5.014;

use strict;
use warnings;

use registry;
use routines;

use Data::Object::Space;

use Scalar::Util ();

# FUNCTIONS

# returns the json-schema-based engine ruleset
fun ruleset() {

  state $ruleset = require Rewire::Ruleset;
}

# returns context closure
fun context() {
  my $cache = {};

  fun(Str $name, Any $object) {

    return $cache->{$name} = $object if $object;
    return $cache->{$name};
  }
}

# returns context with eager-loaded objects
fun preload(HashRef $servConf, Maybe[CodeRef] $context) {
  $context = context() if !$context;

  if (my $servSpec = $servConf->{services}) {
    for my $name (keys %$servSpec) {
      next if $context->($name);

      my $service = $servSpec->{$name};
      my $lifecycle = $service->{lifecycle};

      next if !$lifecycle;

      if ($lifecycle eq 'eager') {
        $context->($name, reifier($name, $servConf, $context));
      }
    }
  }

  $context;
}

# builds and returns object or value based on spec
fun builder(HashRef $service, Any $argument, Maybe[Object] $construct) {
  my $space = Data::Object::Space->new($service->{package});

  # load declared package
  $space->load;

  # determine how to pass arguments (if any)
  my @arguments = arguments($argument, $service->{argument_as});

  # determine construction
  if (my $builder = $service->{builder}) {
    my $original;

    # inject at last build step unless arguments exist
    my @injectables = @arguments;

    for (my $i=0; $i < @$builder; $i++) {
      my $buildspec = $builder->[$i];
      my $argument = $buildspec->{argument};
      my $argument_as = $buildspec->{argument_as};
      my $return = $buildspec->{return};

      my $result = $construct || $space->package;
      my @arguments = arguments($argument, $argument_as);

      if ($i == $#$builder) {
        # on last build step if no build step arguments
        @arguments = @injectables if not exists $buildspec->{argument};
      }

      if (my $function = $buildspec->{function}) {
        $result = $space->call($function, @arguments);
      }
      elsif (my $method = $buildspec->{method}) {
        $result = $space->call($method, $result, @arguments);
      }
      elsif (my $routine = $buildspec->{routine}) {
        $result = $space->call($routine, $space->package, @arguments);
      }
      else {
        next;
      }

      if ($return eq 'class') {
        $construct = $space->package;
      }
      if ($return eq 'result') {
        $construct = $result;
      }
      if ($return eq 'self') {
        $construct = $original //= $result;
      }
    }
  }
  elsif (my $method = $service->{method}) {
    $construct = $space->package->$method(@arguments);
  }
  elsif (my $function = $service->{function}) {
    $construct = $space->call($function, @arguments);
  }
  elsif (my $routine = $service->{routine}) {
    $construct = $space->call($routine, $space->package, @arguments);
  }
  elsif (my $constructor = $service->{constructor}) {
    $construct = $space->package->$constructor(@arguments);
  }
  else {
    $construct = $space->build(@arguments);
  }

  $construct;
}

# returns invoked object or value based on service name
fun reifier(Str $servName, HashRef $servConf, Maybe[CodeRef] $context) {
  $context = preload($servConf) if !$context;

  my $value;
  my $service;
  my $extended;

  my $servSpec = $servConf->{services};

  # use cached (if any)
  $service = $context->($servName);

  return $service if $service;

  $service = $servSpec->{$servName} or return;

  # extend existing service (if requested)
  if (my $extends = $service->{extends}) {
    $extended = reifier($extends, $servConf, $context);
  }

  # build object or value
  my $arguments = resolver($service->{argument}, $servConf, $context);

  $value = builder($service, $arguments, $extended);

  # determine cachability
  my $lifecycle = $service->{lifecycle};

  if ($lifecycle && $lifecycle eq 'singleton') {
    $context->($servName, $value);
  }

  return $value;
}

# resolves service spec arguments
fun resolver(Any $argsData, HashRef $servConf, Maybe[CodeRef] $context) {
  my $servMeta = $servConf->{metadata};
  my $servSpec = $servConf->{services};

  if (ref $argsData eq 'ARRAY') {
    $argsData = [map resolver($_, $servConf, $context), @$argsData];
  }

  # $metadata
  if (ref $argsData eq 'HASH' && (keys %$argsData) == 1) {
    if ($servMeta && $argsData->{'$metadata'}) {
      $argsData = $servMeta->{$argsData->{'$metadata'}};
    }
  }

  # $envvar
  if (ref $argsData eq 'HASH' && (keys %$argsData) == 1) {
    if (my $envvar = $argsData->{'$envvar'}) {
      if (exists $ENV{$envvar}) {
        $argsData = $ENV{$envvar};
      }
      elsif (exists $ENV{uc($envvar)}) {
        $argsData = $ENV{uc($envvar)};
      }
      else {
        $argsData = undef;
      }
    }
  }

  # $service
  if (ref $argsData eq 'HASH' && (keys %$argsData) == 1) {
    if ($servSpec && $argsData->{'$service'}) {
      $argsData = reifier($argsData->{'$service'}, $servConf, $context);
    }
  }

  if (ref $argsData eq 'HASH' && grep ref, values %$argsData) {
    @$argsData{keys %$argsData} = map resolver($_, $servConf, $context), values %$argsData;
  }

  return $argsData;
}

# returns a list of arguments for object construction
fun arguments(Any $argument, Any $argument_as) {
  my @arguments;

  if ($argument && $argument_as) {
    if ($argument_as eq 'array') {
      if (ref $argument eq 'HASH') {
        @arguments = ([$argument]);
      }
      else {
        @arguments = ($argument);
      }
    }
    if ($argument_as eq 'hashmap') {
      if (ref $argument eq 'ARRAY') {
        @arguments = ({@$argument});
      }
      else {
        @arguments = ($argument);
      }
    }
    if ($argument_as eq 'list') {
      if (ref $argument eq 'ARRAY') {
        @arguments = (@$argument);
      }
      elsif (ref $argument eq 'HASH') {
        @arguments = (%$argument);
      }
      else {
        @arguments = ($argument);
      }
    }
  }
  else {
    @arguments = ($argument) if defined $argument;
  }

  (@arguments);
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Dependency Injection Container for Perl 5",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Rewire",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Rewire",
  "PREREQ_PM" => {
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "Data::Object::Role::Buildable" => "0.03",
    "Data::Object::Role::Proxyable" => "2.03",
    "Data::Object::Space" => "2.06",
    "JSON::Validator" => "3.25",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "Data::Object::Role::Buildable" => "0.03",
    "Data::Object::Role::Proxyable" => "2.03",
    "Data::Object::Space" => "2.06",
    "JSON::Validator" => "3.25",
    "Test::Auto" => "0.12",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Object::Class" => "2.02",
  "Data::Object::ClassHas" => "2.01",
  "Data::Object::Role::Buildable" => "0.03",
  "Data::Object::Role::Proxyable" => "2.03",
  "Data::Object::Space" => "2.06",
  "JSON::Validator" => "3.25",
  "Test::Auto" => "0.12",
  "routines" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
